### PR TITLE
WIP: Make the Avro to Iceberg schema converter similar to the spark Avro t…

### PR DESCRIPTION
**Pull request by johnclara**
…o Spark SchemaConverter

The goal of this PR is to get consistency with spark-avro's SchemaConverters.

For complex union types, they convert it to a struct with optional fields.

https://github.com/apache/spark/blob/master/external/avro/src/main/scala/org/apache/spark/sql/avro/SchemaConverters.scala#L120

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/avro/SchemaToType.java [hunk 1]
```diff
@@ -102,15 +107,47 @@ public Type record(Schema record, List<String> names, List<Type> fieldTypes) {
...
+
+      Set<Type.TypeID> doubleTypes = new HashSet<>();
+      longTypes.add(Type.TypeID.FLOAT);
+      longTypes.add(Type.TypeID.DOUBLE);
+
+      if (options.size() == 1) {
+        return options.get(0);
+      } else if (options.size() == 2 && isOfType(options.get(0), longTypes) && isOfType(options.get(1), longTypes)) {
+        return Types.LongType.get();
+      } else if (options.size() == 2 && isOfType(options.get(0), longTypes) && isOfType(options.get(1), doubleTypes)) {
```
Comment (johnclara) [inline core/src/main/java/org/apache/iceberg/avro/SchemaToType.java:138]: The changes for size 1 and Long/Doubles are just trying to line up with the spark-avro SchemaConverter. these cna be removed

File: core/src/main/java/org/apache/iceberg/avro/SchemaToType.java [hunk 2]
```diff
@@ -21,7 +21,12 @@
 
 import com.google.common.base.Preconditions;
 import com.google.common.collect.Lists;
+
+import java.util.HashSet;
+import java.util.LinkedList;
 import java.util.List;
+import java.util.Set;
+
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/avro/SchemaToType.java:29]: We don't add blank lines in imports.

File: core/src/main/java/org/apache/iceberg/avro/SchemaToType.java [hunk 3]
```diff
@@ -102,15 +107,47 @@ public Type record(Schema record, List<String> names, List<Type> fieldTypes) {
...
+      }
     } else {
-      return options.get(0);
+      Set<Type.TypeID> longTypes = new HashSet<>();
+      longTypes.add(Type.TypeID.INTEGER);
+      longTypes.add(Type.TypeID.LONG);
+
+      Set<Type.TypeID> doubleTypes = new HashSet<>();
+      longTypes.add(Type.TypeID.FLOAT);
+      longTypes.add(Type.TypeID.DOUBLE);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/avro/SchemaToType.java:132]: Looks like these should be static final constants?

**Comment by rdblue** [comment]
What is the use case you're working on? Right now, there is no way to read Avro data files if they contain non-option unions and I'm not sure that we want to add that. I'm curious what you intend to use this for.

**Comment by johnclara** [comment]
We're trying to make a table with a union type which has multiple fields. 

Our use case is for CDC of our DB. The changed fields are extracted as Map[String, Object] where Object is a large union type.

I think I'll add a different layer using a custom AvroSchemaVisitor to do this transformation before passing the Schema to the SchemaToType transformer. That way this doesn't have to be merged up stream.

Thanks for taking a look!

**Comment by sudssf** [comment]
@rdblue  have map in avro schema with value as union of different types. spark-avro handles this correctly. looks like iceberg-avro does not support this behaviour
we are tying to figure out if we can add support similar to spark-avro

```
{
         "type": "map",
          "values": [
            "null",
            "int",
            "long",
            "string"
          ]
 }````

**Comment by rdsr** [comment]
I think this is similar to the issue I filed sometime back https://github.com/apache/incubator-iceberg/issues/189 .  @shardulm94  was also playing around with prototypes on how to solve it.

**Comment by sudssf** [comment]
thanks @rdsr will take a look at the issue

**Pull request closed without merging**